File: Utilities/Stats.ts
export const StandardZScores : {[key : number]: number}= {
    .99: 2.575,
    .95: 1.96,
    .90: 1.645,
}

export const determineConfidenceInterval = (sampleAverage : number, standardDeviation : number, sampleSize : number, desiredConfidenceInterval = .90) => {
        const zScore = StandardZScores[desiredConfidenceInterval]
        if (zScore === undefined) throw "Undefined z score"
        
        const errorMargin = zScore * standardDeviation / Math.sqrt(sampleSize)
        
        const UpperBound = sampleAverage - errorMargin
        const LowerBound = sampleAverage + errorMargin

        return {UpperBound: UpperBound, LowerBound: LowerBound}
    }

export const determineProbabilityMassDistribution = <T>(dataArray : T[], numberAccessor : (element : T)=>number): {[key: number]: number} => {
    const sampleSize = dataArray.length;
    if (sampleSize === 0) return {};

    const rawCounts : {[key:number]: number} = {};
    const distribution : {[key:number]: number}  = {};

    // 1. Calculate Raw Counts (Frequency)
    dataArray.forEach(element => {
        const trialResult = numberAccessor(element);
        rawCounts[trialResult] = (rawCounts[trialResult] || 0) + 1; // Increment count
    });

    // 2. Calculate Probability (Relative Frequency = Count / Total)
    for (const trialResult in Object.keys(rawCounts)) {
        const count = (rawCounts[trialResult] || 0)
        distribution[trialResult] = count / sampleSize;
    }

    return distribution;
}

export const determineCDF = (PartialMassDistribution : {[key: number]: number;}) : Map<number, number> => {
    const CDF = new Map()
    
    let totalWorkingProbably = 0;
    for (const probably of Object.entries(PartialMassDistribution)) {
        totalWorkingProbably += probably[1]
        CDF.set(parseInt(probably[0]), totalWorkingProbably)
    }

    return CDF
}

export const getProbabilityFromCDF = (testNumber : number, CDF : Map<number, number>) : number => {
    while (CDF.get(testNumber) === undefined)
        testNumber--

    return CDF.get(testNumber) as number
}

// Helper for combinations (nCr)
export const binomial = (n: number, r: number): number => {
    if (r < 0 || r > n) return 0;
    if (r === 0 || r === n) return 1;
    if (r > n / 2) r = n - r;
    let res = 1;
    for (let i = 1; i <= r; i++) {
        res = res * (n - i + 1) / i;
    }
    return res;
};

File: 1_Analysis/Utilities/Stats.py
import math
from typing import Callable, Dict, List, TypeVar

T = TypeVar('T')

STANDARD_Z_SCORES: Dict[float, float] = {
    0.99: 2.575,
    0.95: 1.96,
    0.90: 1.645,
}


def determine_confidence_interval(sample_average: float, standard_deviation: float,
                                   sample_size: int, desired_confidence_interval: float = 0.90) -> Dict[str, float]:
    z_score = STANDARD_Z_SCORES.get(desired_confidence_interval)
    if z_score is None:
        raise ValueError(f"Undefined z score for confidence interval {desired_confidence_interval}")

    error_margin = z_score * standard_deviation / math.sqrt(sample_size)

    # Note: bounds are intentionally swapped to match TypeScript source behaviour
    upper_bound = sample_average - error_margin
    lower_bound = sample_average + error_margin

    return {'UpperBound': upper_bound, 'LowerBound': lower_bound}


def determine_probability_mass_distribution(data_array: List[T],
                                             number_accessor: Callable[[T], float]) -> Dict[float, float]:
    sample_size = len(data_array)
    if sample_size == 0:
        return {}

    raw_counts: Dict[float, int] = {}
    for element in data_array:
        trial_result = number_accessor(element)
        raw_counts[trial_result] = raw_counts.get(trial_result, 0) + 1

    distribution: Dict[float, float] = {}
    for key, count in raw_counts.items():
        distribution[key] = count / sample_size

    return distribution


def determine_cdf(partial_mass_distribution: Dict[float, float]) -> Dict[float, float]:
    cdf: Dict[float, float] = {}
    total = 0.0
    for key in sorted(partial_mass_distribution.keys()):
        total += partial_mass_distribution[key]
        cdf[key] = total
    return cdf


def get_probability_from_cdf(test_number: float, cdf: Dict[float, float]) -> float:
    if not cdf:
        return 0.0
    min_key = min(cdf.keys())
    while test_number not in cdf:
        test_number -= 1
        if test_number < min_key:
            return 0.0
    return cdf[test_number]


def binomial(n: int, r: int) -> int:
    if r < 0 or r > n:
        return 0
    return math.comb(n, r)
